ClassifAI NUX

iamdharmesh opened the pull request:
### Description of the Change
PR Adds new user onboarding and updates existing settings UI.
<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #30 

#### Screenshots:
**Activation Notice:**
<img width="1339" alt="image" src="https://user-images.githubusercontent.com/10613171/229032892-4eaca467-f437-4275-8b1b-963e5a8dfabd.png">

**ClassifAI Step 1:** 
![screencapture-classifai-local-wp-admin-admin-php-2023-04-28-20_06_33](https://user-images.githubusercontent.com/10613171/235177236-4619bd3a-6fd6-4626-998b-0f0505271d2d.png)


**ClassifAI Step 2:** 
![Screenshot 2023-03-31 at 11 16 01 AM](https://user-images.githubusercontent.com/10613171/229034308-c820255a-20d3-4085-9271-60121a5b65ff.png)

**ClassifAI Step 3:** 
![Screenshot 2023-03-31 at 11 14 31 AM](https://user-images.githubusercontent.com/10613171/229034289-f2b22b6b-e334-48c2-86ce-796d8a3ec441.png)

**Setup Summary:** 
<img width="1508" alt="Screenshot 2023-04-28 at 8 04 12 PM" src="https://user-images.githubusercontent.com/10613171/235177332-d02edfc6-a657-4bc5-abdb-ed022503c9c8.png">


**Classifai Settings:**
![Screenshot 2023-03-31 at 11 18 55 AM](https://user-images.githubusercontent.com/10613171/229034300-1cec0888-b439-40ce-948d-34d1ab49af0a.png)
![Screenshot 2023-03-31 at 11 19 05 AM](https://user-images.githubusercontent.com/10613171/229034294-7249fe96-9624-4194-a115-65153481ca8a.png)

### Reviewer Notes:
- Placeholder images are added in step 1 and the summary step. it will be replaced with actual images once we have images available. I have made PR ready for review to unblock the process here. (@jeffpaul could please help to get images for the same)
- E2E tests are not added as part of this PR, I will add it in the follow-up PR.

### How to test the Change
1. Install and activate the plugin.
2. Verify that the notice for start setup is there.
3. Start ClassifAI set up by clicking the button in the activation notice.
4. Go through each step of setup and make sure everything works fine
5. Go to ClassifAI settings and check the new UI and make sure everything works fine.

### Changelog Entry
> Added - Classifai Setup onboarding and UI improvements.

### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @ryanwelcher @jeffpaul @iamdharmesh 


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

iamdharmesh committed:
Added plugin activation notice for start setup

iamdharmesh committed:
Added initial classifai header.

iamdharmesh committed:
Added onboarding page and setup steps.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/30

iamdharmesh committed:
Added onboarding step 1.

iamdharmesh committed:
Added Recommended content in enable features step.

iamdharmesh committed:
fixed VIPCS

iamdharmesh committed:
Added register classifAI step.

iamdharmesh committed:
Updated step value of hidden input.

iamdharmesh committed:
Added onboarding step 3.

iamdharmesh committed:
Updated admin js for support onboarding form.

iamdharmesh committed:
Added handling for other providers in step 3.

iamdharmesh committed:
Update link to classifiAI setup in activation notice.

iamdharmesh committed:
Keep settings selected for enabled features.

iamdharmesh committed:
Added final onboarding step.

iamdharmesh committed:
Wording updates.

iamdharmesh committed:
Handled skip setup.

iamdharmesh committed:
Prevent users to accessing wizard steps directly.

iamdharmesh committed:
CSS improvements.

iamdharmesh committed:
Update feature option save logic.

iamdharmesh committed:
Don't display activation notice if onboarding is already completed.

iamdharmesh committed:
Fixed js formatting.

iamdharmesh committed:
Added header control links.

iamdharmesh committed:
Update existing setting screens.

iamdharmesh committed:
Updated checkbox toggle UI

iamdharmesh committed:
Updated completed step UI

iamdharmesh committed:
Code improvements.

iamdharmesh committed:
Utilize `render_classifai_setup_settings` in step 2 rendering

iamdharmesh committed:
Rename $page variable to $classifai_page

iamdharmesh committed:
Removed unwanted js file.

iamdharmesh committed:
Eslint fixes.

iamdharmesh committed:
Added default features to be enabled.

iamdharmesh committed:
Fixed PHP notice.

Sidsector9 commented on includes/Classifai/Admin/templates/classifai-header.php:
```diff
@@ -0,0 +1,79 @@
+					<span class="dashicons dashicons-editor-help"></span>
+					<span class="control-item-text"><?php esc_html_e( 'Help', 'classifai' ); ?></span>
+				</a>
+				<template id="help-menu-template">
```
👍

Sidsector9 commented on includes/Classifai/Admin/templates/onboarding-step-four.php:
```diff
@@ -0,0 +1,148 @@
+						<ul>
+							<?php
+							$image_features = array(
+								'image_captions' => __( 'Automatically add alt-text to Images', 'classifai' ),
```
"Images" can be images (lower text)

Sidsector9 commented on includes/Classifai/Admin/templates/onboarding-step-one.php:
```diff
@@ -0,0 +1,156 @@
+					<ul>
+						<?php
+						$image_features = array(
+							'image_captions' => __( 'Automatically add alt-text to Images', 'classifai' ),
```
"Images" can be "images"

Sidsector9 reviewed (dismissed):
@iamdharmesh this is really good work 🎊  I tested the flow on a fresh installation and it tested well. 

As for code review I left some really minor changes.

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-four.php:
```diff
@@ -0,0 +1,148 @@
+use function Classifai\get_post_types_for_language_settings;
+
+$onboarding_options   = get_option( 'classifai_onboarding_options', array() );
+$enabled_features     = isset( $onboarding_options['enabled_features'] ) ? $onboarding_options['enabled_features'] : array();
```
Can simplify this:
```suggestion
$enabled_features     =  $onboarding_options['enabled_features'] ?? array();
```

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-four.php:
```diff
@@ -0,0 +1,148 @@
+
+$onboarding_options   = get_option( 'classifai_onboarding_options', array() );
+$enabled_features     = isset( $onboarding_options['enabled_features'] ) ? $onboarding_options['enabled_features'] : array();
+$configured_providers = isset( $onboarding_options['configured_providers'] ) ? $onboarding_options['configured_providers'] : array();
```
```suggestion
$configured_providers = $onboarding_options['configured_providers'] ?? array();
```

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-four.php:
```diff
@@ -0,0 +1,148 @@
+						<ul>
+							<?php
+							$types    = get_post_types_for_language_settings();
+							$features = isset( $enabled_features['language']['classify'] ) ? $enabled_features['language']['classify'] : array();
```
```suggestion
							$features = $enabled_features['language']['classify'] ?? array();
```

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-four.php:
```diff
@@ -0,0 +1,148 @@
+								<li class="classifai-enable-feature">
+									<span class="dashicons dashicons-yes-alt"></span>
+									<label class="classifai-feature-text">
+										<?php esc_html_e( 'Automatic excerpt generation', 'classifai' ); ?>
```
Since we don't actually automatically generate excerpts (you have to click a button to generate it) I think we may want to change the wording here a bit:
```suggestion
										<?php esc_html_e( 'Excerpt generation', 'classifai' ); ?>
```

iamdharmesh replied:
Thanks for adding note for this.

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-four.php:
```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+ * Step-1 template for ClassifAI Onboarding.
```
```suggestion
 * Step-4 template for ClassifAI Onboarding.
```

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-one.php:
```diff
@@ -0,0 +1,156 @@
+		'image_tags'     => 'yes',
+	),
+);
+$enabled_features   = isset( $onboarding_options['enabled_features'] ) ? $onboarding_options['enabled_features'] : $default_enabled;
```
```suggestion
$enabled_features   = $onboarding_options['enabled_features'] ?? $default_enabled;
```

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-one.php:
```diff
@@ -0,0 +1,156 @@
+					<ul>
+						<?php
+						$types    = get_post_types_for_language_settings();
+						$features = isset( $enabled_features['language']['classify'] ) ? $enabled_features['language']['classify'] : array();
```
```suggestion
						$features = $enabled_features['language']['classify'] ?? array();
```

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-one.php:
```diff
@@ -0,0 +1,156 @@
+						$types    = get_post_types_for_language_settings();
+						$features = isset( $enabled_features['language']['classify'] ) ? $enabled_features['language']['classify'] : array();
+						foreach ( $types as $posttype ) {
+							$checked = isset( $features[ $posttype->name ] ) ? $features[ $posttype->name ] : '';
```
```suggestion
							$checked = $features[ $posttype->name ] ?? '';
```

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-one.php:
```diff
@@ -0,0 +1,156 @@
+						<li class="classifai-enable-feature">
+							<label class="classifai-toggle">
+								<span class="classifai-feature-text">
+									<?php esc_html_e( 'Automatic excerpt generation', 'classifai' ); ?>
```
```suggestion
									<?php esc_html_e( 'Excerpt generation', 'classifai' ); ?>
```

dkotter commented on includes/Classifai/Admin/templates/onboarding-step-two.php:
```diff
@@ -0,0 +1,42 @@
+<?php
+/**
+ * Step-1 template for ClassifAI Onboarding.
```
```suggestion
 * Step-2 template for ClassifAI Onboarding.
```

dkotter commented on includes/Classifai/Admin/Onboarding.php:
```diff
@@ -0,0 +1,569 @@
+	public function __construct() {
+		$this->setup_url = admin_url( 'admin.php?page=classifai_setup' );
+
+		add_action( 'admin_menu', [ $this, 'register_setup_page' ] );
```
Not a blocker necessarily but I prefer to not add hooks within a constructor. Typical approach is an `init` or `setup` method that holds these hooks and wherever we instantiate a new class we can call that method:

```php
$onboarding = new Classifai\Admin\Onboarding();
$onboarding->init();
```

iamdharmesh replied:
Thanks for adding a note for this. I have updated the code accordingly.

dkotter commented on includes/Classifai/Admin/Onboarding.php:
```diff
@@ -0,0 +1,569 @@
+		}
+
+		$onboarding_options   = self::get_onboarding_options();
+		$configured_providers = isset( $onboarding_options['configured_providers'] ) ? $onboarding_options['configured_providers'] : array();
```
```suggestion
		$configured_providers = $onboarding_options['configured_providers'] ?? array();
```

dkotter commented on includes/Classifai/Admin/Onboarding.php:
```diff
@@ -0,0 +1,569 @@
+		$providers          = self::get_setup_providers();
+		$enabled_providers  = array();
+		$onboarding_options = self::get_onboarding_options();
+		$enabled_features   = isset( $onboarding_options['enabled_features'] ) ? $onboarding_options['enabled_features'] : array();
```
```suggestion
		$enabled_features   = $onboarding_options['enabled_features'] ?? array();
```

dkotter commented:
I need to test this out a bit more but at a first glance, this looks way better! I've done a first pass code review and left a handful of comments.

I do have a few other things that came to mind that may be worth discussing here:

- Wondering if there's a better way to easily connect providers and the features they provide into this flow? For instance, if a new provider gets added or a new feature gets added to an existing provider, there seems to be multiple places now that will need updated in order to get that into this onboarding flow. Would be great if we could find a way to simplify this. Maybe some sort of method in our base `Provider` class that each of our providers can then extend, detailing out what features that provider has, which can then be pulled into the onboarding? Just not sure I feel great about having a bunch of information hardcoded into the onboarding flow and the maintenance of that long term
- Going along with the above, seems some duplication around saving of some options for each provider, like API keys and URLs. Again just thinking if the structure of that data ever changes, we're now managing that in two separate locations. May be someway to set this structure up in each of our `Provider` classes and pull into the onboarding flow
- Also noticed there's a decent amount of duplicate code (particularly HTML) in a few of the onboarding steps. Probably fine but may be something we can look to simplify
- And not to throw a huge wrench into the current approach but did come to my mind that this would be a great opportunity to build out a really nice, interactive UI using something like React and all the various components that WordPress comes with now. Not saying we need to go that path but worth talking about (and may simplify some of the duplicate HTML mentioned above)

ankitguptaindia commented:
Hello @iamdharmesh thanks for this PR. New enhancement looks great. I checked these new changes briefly and found some areas where we can improve more.

1. After activation of the plugin, the onboarding notice appears and disappears after the page refreshes or navigating to any other page. It should be available in the header so user can easily start setting up things. 

Refer - 

https://user-images.githubusercontent.com/6384299/229272357-a0b92bc4-abdf-4a15-a165-de3e2169ed0d.mp4

---

2. Newly added setup screen doesn't have a Page Title. 


Refer - 

![Screenshot 2023-04-01 at 1 04 42 PM](https://user-images.githubusercontent.com/6384299/229272515-b9628d80-c10b-4e53-9f8d-66ed5423cbb6.png)

---

3. On the Second Step 'Register ClassifAI' we have two options, 1. Skip for now, 2. Register

When the user select 'Register' option it shows fields to add Email and Key but user can move ahead without filling these fields and header status appear as checked. 
It should not allow user to move ahead without filling email/key as if a user doesn't have the email/key then user will select option  'Skip for now'.


https://user-images.githubusercontent.com/6384299/229273128-2f34e275-1a11-4657-acb4-b4df9396019d.mp4

---

QA- WIP. Need to discuss some concerns related to this PR scope. Will resume QA after that.

iamdharmesh committed:
PR feedback updates.

iamdharmesh committed:
Migrate features options to provider class.

iamdharmesh committed:
Handle step submission with common function.

iamdharmesh committed:
Get default feaure from default settings.

iamdharmesh committed:
Added onboarding header and footer templates.

iamdharmesh commented:
Thanks for the detailed review @dkotter,

> Wondering if there's a better way to easily connect providers and the features they provide into this flow? For instance, if a new provider gets added or a new feature gets added to an existing provider, there seems to be multiple places now that will need updated in order to get that into this onboarding flow. Would be great if we could find a way to simplify this. Maybe some sort of method in our base Provider class that each of our providers can then extend, detailing out what features that provider has, which can then be pulled into the onboarding? Just not sure I feel great about having a bunch of information hardcoded into the onboarding flow and the maintenance of that long term

Agree, I already utilized the `Provider` class to render forms for setting up providers. but features and providers were not well connected. I have updated the code and now we are pulling features from the `Provider` class. I have added new class property  `onboarding_options` and it will handle all onboarding-related stuff from the `Provider` class now. Please help to review the updated code once and let me know if we need any further changes here.

> Going along with the above, seems some duplication around saving of some options for each provider, like API keys and URLs. Again just thinking if the structure of that data ever changes, we're now managing that in two separate locations. May be someway to set this structure up in each of our Provider classes and pull into the onboarding flow

Thanks for raising this, This is also handled with the previous issue.

> Also noticed there's a decent amount of duplicate code (particularly HTML) in a few of the onboarding steps. Probably fine but may be something we can look to simplify

Yes, I have added header/footer templates for onboarding and not we don't have much duplicate HTML and other code. Please let me know if I missed anything here which is still need to be addressed.

> And not to throw a huge wrench into the current approach but did come to my mind that this would be a great opportunity to build out a really nice, interactive UI using something like React and all the various components that WordPress comes with now. Not saying we need to go that path but worth talking about (and may simplify some of the duplicate HTML mentioned above)

Yes, As an initial thought, I also considered going with React. But after considering the setting up Providers form, APIs and efforts, I decided to go with this approach in order to re-utilize the provider class functionality. However, React is also a great option here, please let me know if you think we should proceed with it here and discard the approach we have taken here.


@ankitguptaindia Thanks for reviewing PR, I haven't addressed the issue reported by you yet. Will fix it once the review is done from your side.

Thanks.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/30

iamdharmesh committed:
Made services menu dynamic.

iamdharmesh committed:
Fix configured features screen.

iamdharmesh committed:
Added onboarding support for DALL·E

iamdharmesh committed:
Fix DALL-E features key.

dkotter commented:
Code here is looking good. I've not done an in-depth round of testing yet though.

mehidi258 commented on includes/Classifai/Admin/Notifications.php:
```diff
@@ -48,10 +48,25 @@ public function maybe_render_notices() {
+			<div data-notice="plugin-activation" class="notice notice-success is-dismissible">
+				<div id="classifai-activation-notice">
+					<div class="classifai-logo">
+						<img src="<?php echo esc_url( CLASSIFAI_PLUGIN_URL . 'assets/img/classifai.png' ); ?>" alt="ClassifAI" />
```
escaping function missing on alt attribute

dkotter replied:
Because we are just outputting a hardcoded string (no translation or anything) there is no need to escape this

mehidi258 replied:
Thanks for confirming, @dkotter I thought when you are echoing something we should escape.

mehidi258 commented on includes/Classifai/Admin/templates/classifai-header.php:
```diff
@@ -0,0 +1,79 @@
+	<div class="classifai-header-layout">
+		<div id="classifai-branding">
+			<div id="classifai-logo">
+				<img src="<?php echo esc_url( CLASSIFAI_PLUGIN_URL . 'assets/img/classifai.png' ); ?>" alt="ClassifAI" />
```
escaping function missing on alt attribute

dkotter replied:
Same here, this is just a hardcoded string so no need to escape

mehidi258 commented on includes/Classifai/Admin/templates/classifai-header.php:
```diff
@@ -0,0 +1,79 @@
+				</a>
+				<template id="help-menu-template">
+					<div class="classifai-help-menu">
+						<a class="classifai-help-menu__menu-item" target="_blank" href="https://github.com/10up/classifai#frequently-asked-questions"><?php esc_html_e( 'FAQs', 'classifai' ); ?></a>
```
Please add `rel="noopener noreferrer"` since it's external link and we are setting target="_blank"

mehidi258 commented on includes/Classifai/Admin/templates/classifai-header.php:
```diff
@@ -0,0 +1,79 @@
+				<template id="help-menu-template">
+					<div class="classifai-help-menu">
+						<a class="classifai-help-menu__menu-item" target="_blank" href="https://github.com/10up/classifai#frequently-asked-questions"><?php esc_html_e( 'FAQs', 'classifai' ); ?></a>
+						<a class="classifai-help-menu__menu-item" target="_blank"  href="https://github.com/10up/classifai/issues/new/choose"><?php esc_html_e( 'Report issue/enhancement', 'classifai' ); ?></a>
```
Same `rel="noopener noreferrer"` here

mehidi258 commented on includes/Classifai/Admin/templates/classifai-header.php:
```diff
@@ -0,0 +1,79 @@
+	<?php
+	foreach ( $classifai_header_menu as $key => $value ) {
+		?>
+		<a href="<?php echo esc_url( admin_url( 'admin.php?page=' . $key ) ); ?>" class="nav-tab <?php echo ( $classifai_page === $key ) ? 'nav-tab-active' : ''; ?>">
```
Add escaping function here `<?php echo ( $classifai_page === $key ) ? 'nav-tab-active' : ''; ?>`

dkotter replied:
Same here, because we end up just outputting either `nav-tab-active` or a blank string, no need to escape that

mehidi258 commented on includes/Classifai/Admin/templates/onboarding-step-four.php:
```diff
@@ -0,0 +1,80 @@
+require_once 'onboarding-header.php';
+?>
+
+<h1 class="classifai-setup-title">
```
Since h1 is already exist, lets change it to some other tag like h2

iamdharmesh replied:
Nice catch!

mehidi258 reviewed (commented):
Overall looks great to me, Though i have added few very nitpicky improvements from FE perspective which are great to have.

iamdharmesh committed:
PR feedback improvements.

iamdharmesh commented:
Thanks for the review @mehidi258. I have updated the code accordingly. Could you please help to review it once?

mehidi258 commented:
> Thanks for the review @mehidi258. I have updated the code accordingly. Could you please help to review it once?

@iamdharmesh Thanks for resolving those, Looks great to me. Though regarding adding escaping functions @dkotter mentioned that those escaping are not needed. Sorry for flagging those, feel free to revert those escpaing changes

iamdharmesh committed:
Move classifai menu under Tools.

iamdharmesh committed:
Update tests.

iamdharmesh committed:
Added plugin action links.

iamdharmesh committed:
Skip registration step if it is already configured.

iamdharmesh committed:
Always show setup link.

iamdharmesh committed:
- Enable all features for new user, show X before not configured feature.

iamdharmesh committed:
Add disable unchecked feature functionality.

iamdharmesh committed:
Prevent showing the default WordPress "Plugin Activated" notice.

iamdharmesh commented:
Hi @jeffpaul @dkotter, I have updated PR for the points we discussed during the demo. Could you please help to review it once? Also, Please let me know if any wording updates are needed anywhere. 

**Header:**
- [x] Keep the "Set up" link always there. (even after setup is complete)

**Step - 1 (Enable features):**
- [x] **Default Enabled features:**
   For new users: ALL features are enabled by default except any CPTs for Language Processing
   For existing users: Only completely configured features should be enabled by default

**Step - 2 (Registration):**
- [x] Skip it if it is already configured

**Step - 4 (summary page):**
- [x] Add not configured features as well with (x) icon before the feature name

**Plugins list page:**
- [x] Add setup and settings link in the plugins list page

**General:**
- [x] Move the ClassifAI menu under the tools menu and remove the top-level menu
- [x] Page title for “ClassifAI Setup” page
- [x] Disable WordPress "Plugin Activated" notice.

Thanks

dkotter commented on classifai.php:
```diff
@@ -65,6 +65,7 @@ function classifai_define( $name, $value ) {
 }
 
 require_once __DIR__ . '/config.php';
+classifai_define( 'CLASSIFAI_PLUGIN_BASENAME', plugin_basename( __FILE__ ) );
```
Feels like this belongs in the `config.php` file, unless there's a reason it can't be there

jeffpaul replied:
Probably just an oddity with ClassifAI not including those bits in the classifai.php file whereas most our other plugins have them in the main plugin file 🤷🏼

iamdharmesh replied:
@dkotter I have kept this in the main plugin file due to `plugin_basename( __FILE__ ) )` because moving this to config.php will give us `config.php`  in plugin basename

iamdharmesh committed:
Added images to welcome and summary step.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/30
